### ISSUE TITLE
Decimal::from_parts() taking a u32 scale can trample rest of Decimal.flags

### ISSUE THREAD

**Issue by duchainer** [paupino/rust-decimal#319]
Hi, this is my first rust issue, so feel free to comment on what could be better/clearer, I'll try to re-edit this post for clarity. ^^ 

# Summary
## Problem
The `scale` parameter of `from_parts` is programmatically unconstrained enough that, if big enough, it can squash the rest of the flags including the negative bit.
## Solution
- Modulo 29 the scale in flags() to keep the same api
- Make scale constrained type by 0..29 range so that the compiler can type check it at compile time, the from_part() return can stay a non-wrapped Decimal that way.

# Long form

While trying to generate multiple fake rust_decimal::Decimal with the [fake-rs](https://github.com/cksac/fake-rs) crate, I spotted an issue with the scale parameter in Decimal::from_parts(). When out of bound the prescribed bounds of `0..29`, it starts to trample on the rest of flags, including the negative bit. This is not a bit issue as long as it doesn't reach the negative bit, but if/when `flags` get added some more flags, it could cause some weird bugs for the library user.

[As a comparison, this is the current distilled result of that ](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=%2F%2F%20Number%20of%20bits%20scale%20is%20shifted%20by.%0Aconst%20SCALE_SHIFT%3A%20u32%20%3D%2016%3B%0A%2F%2F%20Number%20of%20bits%20sign%20is%20shifted%20by.%0Aconst%20SIGN_SHIFT%3A%20u32%20%3D%2031%3B%0A%0Aconst%20fn%20flags(neg%3A%20bool%2C%20scale%3A%20u32)%20-%3E%20u32%20%7B%0A%20%20%20%20(scale%20%3C%3C%20SCALE_SHIFT)%20%7C%20((neg%20as%20u32)%20%3C%3C%20SIGN_SHIFT)%0A%7D%0A%0Afn%20main()%20%7B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(true%2C%20u32%3A%3AMAX))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(false%2C%20u32%3A%3AMAX))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(true%2C%2029u32))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(false%2C%2029u32))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(true%2C%2028u32))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(false%2C%2028u32))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(true%2C%200u32))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(false%2C%200u32))%3B%0A%20%20%20%20assert_eq!(0x800F0000%2C%20flags(true%2C%20u32%3A%3AMAX))%3B%0A%20%20%20%20assert_eq!(0xF0000%2C%20flags(false%2C%20u32%3A%3AMAX))%3B%0A%20%20%20%20assert_eq!(0x80000000%2C%20flags(true%2C%2029u32))%3B%0A%20%20%20%20assert_eq!(0x0%2C%20flags(false%2C%2029u32))%3B%0A%20%20%20%20assert_eq!(0x801C0000%2C%20flags(true%2C%2028u32))%3B%0A%20%20%20%20assert_eq!(0x1C0000%2C%20flags(false%2C%2028u32))%3B%0A%20%20%20%20assert_eq!(0x80000000%2C%20flags(true%2C%200u32))%3B%0A%20%20%20%20assert_eq!(0x0%2C%20flags(false%2C%200u32))%3B%0A%7D): 
```
// Number of bits scale is shifted by.
const SCALE_SHIFT: u32 = 16;
// Number of bits sign is shifted by.
const SIGN_SHIFT: u32 = 31;

const fn flags(neg: bool, scale: u32) -> u32 {
    (scale << SCALE_SHIFT) | ((neg as u32) << SIGN_SHIFT)
}

fn main() {
    println!("{:X}", flags(true, u32::MAX));
    println!("{:X}", flags(false, u32::MAX));
    println!("{:X}", flags(true, 29u32));
    println!("{:X}", flags(false, 29u32));
    println!("{:X}", flags(true, 28u32));
    println!("{:X}", flags(false, 28u32));
    println!("{:X}", flags(true, 0u32));
    println!("{:X}", flags(false, 0u32));
    assert_eq!(0x800F0000, flags(true, u32::MAX));
    assert_eq!(0xF0000, flags(false, u32::MAX));
    assert_eq!(0x80000000, flags(true, 29u32));
    assert_eq!(0x0, flags(false, 29u32));
    assert_eq!(0x801C0000, flags(true, 28u32));
    assert_eq!(0x1C0000, flags(false, 28u32));
    assert_eq!(0x80000000, flags(true, 0u32));
    assert_eq!(0x0, flags(false, 0u32));
}
```
[A quick fix could be to modulo 29 the scale in flags to enforce a scale between 0 and 28 inclusively without changing the current api](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=%2F%2F%20Number%20of%20bits%20scale%20is%20shifted%20by.%0Aconst%20SCALE_SHIFT%3A%20u32%20%3D%2016%3B%0A%2F%2F%20Number%20of%20bits%20sign%20is%20shifted%20by.%0Aconst%20SIGN_SHIFT%3A%20u32%20%3D%2031%3B%0A%0Aconst%20fn%20flags(neg%3A%20bool%2C%20scale%3A%20u32)%20-%3E%20u32%20%7B%0A%20%20%20%20((scale%20%25%2029)%20%3C%3C%20SCALE_SHIFT)%20%7C%20((neg%20as%20u32)%20%3C%3C%20SIGN_SHIFT)%0A%7D%0A%0Afn%20main()%20%7B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(true%2C%20u32%3A%3AMAX))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(false%2C%20u32%3A%3AMAX))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(true%2C%2029u32))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(false%2C%2029u32))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(true%2C%2028u32))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(false%2C%2028u32))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(true%2C%200u32))%3B%0A%20%20%20%20println!(%22%7B%3AX%7D%22%2C%20flags(false%2C%200u32))%3B%0A%20%20%20%20assert_eq!(0x800F0000%2C%20flags(true%2C%20u32%3A%3AMAX))%3B%0A%20%20%20%20assert_eq!(0xF0000%2C%20flags(false%2C%20u32%3A%3AMAX))%3B%0A%20%20%20%20assert_eq!(0x80000000%2C%20flags(true%2C%2029u32))%3B%0A%20%20%20%20assert_eq!(0x0%2C%20flags(false%2C%2029u32))%3B%0A%20%20%20%20assert_eq!(0x801C0000%2C%20flags(true%2C%2028u32))%3B%0A%20%20%20%20assert_eq!(0x1C0000%2C%20flags(false%2C%2028u32))%3B%0A%20%20%20%20assert_eq!(0x80000000%2C%20flags(true%2C%200u32))%3B%0A%20%20%20%20assert_eq!(0x0%2C%20flags(false%2C%200u32))%3B%0A%7D%0A): 
```
const fn flags(neg: bool, scale: u32) -> u32 {
    ((scale % 29) << SCALE_SHIFT) | ((neg as u32) << SIGN_SHIFT)
}
```

If this is considered useful, I'll submit a PR to fix it.

**Comment by paupino** [comment]
Thanks for the great write up!

Thinking this through a little bit: `flags` is currently intended to be an inline helper function which bases itself upon assumptions that the data structure provides. Assumptions like this are used within `Decimal` at the moment, largely in operations, to help extrapolate performance. While adding an additional check within `flags` would prevent any future problems, my concern is that it's used everywhere. That being said... `flags` is usually only used to construct final state, so may be more of a moot concern however I think for now it may be better to be consistent with other public API calls using scale and constrain it on entry. Identical to [this](https://github.com/paupino/rust-decimal/blob/79237c2e1ecc8f642826d3078c764c87110a009c/src/decimal.rs#L218-L223). The _downside_ is that this now introduces a panic on invalid input - though that is consistent with other APIs. Going back in time (and maybe something I'll consider for v2) I would have made these return `Result`s instead however for now I'm stuck with this design decision.

Personally, I think that a potential panic on entry is probably fairly reasonable given the intention of `from_parts`.  It's primary use case is for macro generation of which I think the check could/would be optimized out. 

Anyway, sorry for this stream of consciousness - I realize this doesn't directly solve your problem, but rather curbs API behavior to prevent it. Does this seem reasonable?

**Comment by duchainer** [comment]
No problem, stream of consciousness are a personal staple too. ^^ 
Also, no rush, this is no big issue on my part, just needing to add a modulo 29 on my end.

After looking at your answer, I think the question is one of undefined behavior when exceeding the MAX_PRECISION, so a panic  is reasonable as a "non-breaking change" as the scale has constraints defined clearly in the doc. Even more so, as `from_i128_with_scale()` uses the same assertion.

**Comment by paupino** [comment]
Fixed in #337. Opted towards keeping it a `const fn` and instead using the suggested modulo approach.
